[SPARK-22239][SQL][Python] Enable grouped aggregate pandas UDFs as window functions with unbounded window frames

icexelloss opened the pull request:
## What changes were proposed in this pull request?
This PR enables using a grouped aggregate pandas UDFs as window functions. The semantics is the same as using SQL aggregation function as window functions.

```
       >>> from pyspark.sql.functions import pandas_udf, PandasUDFType
       >>> from pyspark.sql import Window
       >>> df = spark.createDataFrame(
       ...     [(1, 1.0), (1, 2.0), (2, 3.0), (2, 5.0), (2, 10.0)],
       ...     ("id", "v"))
       >>> @pandas_udf("double", PandasUDFType.GROUPED_AGG)
       ... def mean_udf(v):
       ...     return v.mean()
       >>> w = Window.partitionBy('id')
       >>> df.withColumn('mean_v', mean_udf(df['v']).over(w)).show()
       +---+----+------+
       | id|   v|mean_v|
       +---+----+------+
       |  1| 1.0|   1.5|
       |  1| 2.0|   1.5|
       |  2| 3.0|   6.0|
       |  2| 5.0|   6.0|
       |  2|10.0|   6.0|
       +---+----+------+
```

The scope of this PR is somewhat limited in terms of:
(1) Only supports unbounded window, which acts essentially as group by.
(2) Only supports aggregation functions, not "transform" like window functions (n -> n mapping)

Both of these are left as future work. Especially, (1) needs careful thinking w.r.t. how to pass rolling window data to python efficiently. (2) is a bit easier but does require more changes therefore I think it's better to leave it as a separate PR.

## How was this patch tested?

WindowPandasUDFTests

icexelloss committed:
Intial commit of using grouped agg pandas UDFs as window functions

SparkQA commented:
**[Test build #89416 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89416/testReport)** for PR 21082 at commit [`9fdcfe6`](https://github.com/apache/spark/commit/9fdcfe636350be11b02d0429d8f0f18344f93d10).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2360/
Test PASSed.

SparkQA commented:
**[Test build #89416 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89416/testReport)** for PR 21082 at commit [`9fdcfe6`](https://github.com/apache/spark/commit/9fdcfe636350be11b02d0429d8f0f18344f93d10).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class WindowInPandasExec(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89416/
Test FAILed.

icexelloss committed:
White space

icexelloss committed:
Style fixes

SparkQA commented:
**[Test build #89462 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89462/testReport)** for PR 21082 at commit [`dfdb03f`](https://github.com/apache/spark/commit/dfdb03fc9c19b120a046b45b0864f8f405b0ead5).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2397/
Test PASSed.

SparkQA commented:
**[Test build #89462 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89462/testReport)** for PR 21082 at commit [`dfdb03f`](https://github.com/apache/spark/commit/dfdb03fc9c19b120a046b45b0864f8f405b0ead5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89462/
Test FAILed.

icexelloss committed:
Add more tests. Refactor window function type

SparkQA commented:
**[Test build #89584 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89584/testReport)** for PR 21082 at commit [`a2825bf`](https://github.com/apache/spark/commit/a2825bf124578e4ae80e88f910c19aa8bb1c4d62).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2495/
Test PASSed.

SparkQA commented:
**[Test build #89584 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89584/testReport)** for PR 21082 at commit [`a2825bf`](https://github.com/apache/spark/commit/a2825bf124578e4ae80e88f910c19aa8bb1c4d62).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89584/
Test FAILed.

icexelloss committed:
Revert basicLogicalOperators

SparkQA commented:
**[Test build #89586 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89586/testReport)** for PR 21082 at commit [`67b3dd0`](https://github.com/apache/spark/commit/67b3dd0421f61ac83559261604a346cecfc4dc99).

SparkQA commented:
**[Test build #89586 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89586/testReport)** for PR 21082 at commit [`67b3dd0`](https://github.com/apache/spark/commit/67b3dd0421f61ac83559261604a346cecfc4dc99).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89586/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2496/
Test PASSed.

icexelloss committed:
Clean up; Add more tests

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2500/
Test PASSed.

icexelloss committed:
Add docs

SparkQA commented:
**[Test build #89597 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89597/testReport)** for PR 21082 at commit [`29b0395`](https://github.com/apache/spark/commit/29b03953ea53fb729811ee985b4a2f0a63de1c7f).

icexelloss commented:
cc @BryanCutler @ueshin @HyukjinKwon @viirya 

cc @yhuai because of window related changes.

This PR is ready for review now

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2501/
Test PASSed.

SparkQA commented:
**[Test build #89595 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89595/testReport)** for PR 21082 at commit [`85159b8`](https://github.com/apache/spark/commit/85159b88c4ccf372b4bd5560167765eb2acdda3f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89595/
Test FAILed.

SparkQA commented:
**[Test build #89597 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89597/testReport)** for PR 21082 at commit [`29b0395`](https://github.com/apache/spark/commit/29b03953ea53fb729811ee985b4a2f0a63de1c7f).
 * This patch **fails from timeout after a configured wait of \`300m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89597/
Test FAILed.

HyukjinKwon commented:
retest this please

HyukjinKwon commented:
From a very quick look, the test failure looks global.

SparkQA commented:
**[Test build #89612 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89612/testReport)** for PR 21082 at commit [`29b0395`](https://github.com/apache/spark/commit/29b03953ea53fb729811ee985b4a2f0a63de1c7f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2515/
Test PASSed.

SparkQA commented:
**[Test build #89612 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89612/testReport)** for PR 21082 at commit [`29b0395`](https://github.com/apache/spark/commit/29b03953ea53fb729811ee985b4a2f0a63de1c7f).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89612/
Test FAILed.

viirya commented:
retest this please.

SparkQA commented:
**[Test build #89616 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89616/testReport)** for PR 21082 at commit [`29b0395`](https://github.com/apache/spark/commit/29b03953ea53fb729811ee985b4a2f0a63de1c7f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2517/
Test PASSed.

SparkQA commented:
**[Test build #89616 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89616/testReport)** for PR 21082 at commit [`29b0395`](https://github.com/apache/spark/commit/29b03953ea53fb729811ee985b4a2f0a63de1c7f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89616/
Test FAILed.

icexelloss committed:
Fix tests

icexelloss commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala:
```diff
@@ -149,7 +149,7 @@ class AnalysisErrorSuite extends AnalysisTest {
           SortOrder(UnresolvedAttribute("b"), Ascending) :: Nil,
           UnspecifiedFrame)).as('window)),
-    "not supported within a window function" :: Nil)
+    "does not have any window functions" :: Nil)
```
This is because an early analysis exception is thrown by rule ExtractWindowExpressions

icexelloss commented on python/pyspark/sql/tests.py:
```diff
@@ -5156,6 +5156,15 @@ def test_retain_group_columns(self):
             expected1 = df.groupby(df.id).agg(sum(df.v))
             self.assertPandasEqual(expected1.toPandas(), result1.toPandas())
 
+    def test_array_type(self):
```
This is related, but I figured its shouldn't hurt to add an array test in GroupedAggPandasUDFTests..

SparkQA commented:
**[Test build #89682 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89682/testReport)** for PR 21082 at commit [`657a6a5`](https://github.com/apache/spark/commit/657a6a5ababbf816db8bbd19475b8e3e5f4aa2ae).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2563/
Test PASSed.

SparkQA commented:
**[Test build #89682 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89682/testReport)** for PR 21082 at commit [`657a6a5`](https://github.com/apache/spark/commit/657a6a5ababbf816db8bbd19475b8e3e5f4aa2ae).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89682/
Test FAILed.

icexelloss committed:
Fix tests. Allow window expression without non window function expression after optmizition stage

icexelloss committed:
Trival fix

SparkQA commented:
**[Test build #89693 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89693/testReport)** for PR 21082 at commit [`27158d9`](https://github.com/apache/spark/commit/27158d9873d54de9312db9e2db5c88d430589ade).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2568/
Test PASSed.

SparkQA commented:
**[Test build #89693 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89693/testReport)** for PR 21082 at commit [`27158d9`](https://github.com/apache/spark/commit/27158d9873d54de9312db9e2db5c88d430589ade).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89693/
Test PASSed.

icexelloss commented:
Hey @HyukjinKwon @ueshin @BryanCutler I've fixed the tests and I think the PR is in good shape for review now. Could you please take a look when you have time? Thanks!

HyukjinKwon commented:
Will take a close look soon within this weekend as well.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2621/
Test PASSed.

icexelloss commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -268,3 +269,38 @@ object PhysicalAggregation {
+      }
+
+      val windowFunctionType = windowExpressions.map(WindowFunctionType.functionType)
+        .reduceLeft ( (t1: WindowFunctionType, t2: WindowFunctionType) =>
```
If we want to do this in Analyzer, then we would carry the WindowFunctionType in the logical plan. 

I did it this way to avoid changing the logical node. I am open to add WindowFunctionType to the logical plan though. What do other people think?

HyukjinKwon replied:
(BTW: 
```
.reduceLeft {
  ...
}
```
)

icexelloss commented on python/pyspark/sql/functions.py:
```diff
@@ -2321,7 +2323,30 @@ def pandas_udf(f=None, returnType=None, functionType=None):
+       >>> df = spark.createDataFrame(
+       ...     [(1, 1.0), (1, 2.0), (2, 3.0), (2, 5.0), (2, 10.0)],
+       ...     ("id", "v"))
+       >>> @pandas_udf("double", PandasUDFType.GROUPED_AGG)  # doctest: +SKIP
```
Yes exactly. The idea is that the producer of the UDF can produce a grouped agg udf, such as weighted mean, and the consumer can use the UDF in both groupby and window, similar to how SQL aggregation function work.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -2301,10 +2301,12 @@ def pandas_udf(f=None, returnType=None, functionType=None):
 
-       :class:`ArrayType`, :class:`MapType` and :class:`StructType` are currently not supported as
-       output types.
+       :class:`MapType` and :class:`StructType` are currently not supported as output types.
```
@icexelloss, actually should we keep this note? I think this is matched with https://spark.apache.org/docs/latest/sql-programming-guide.html#supported-sql-types which we documented there and SQLConf.

Probably, just leaving a link could be fine. Removing out is okay to me too.

icexelloss replied:
I am leaning towards keeping this in the API doc and maybe make sql-programming-guide link to this.

I think most user would look for API docs first rather than sql-programming-guide, so it's probably a bit more convenient to have it here?

icexelloss replied:
Track in: https://issues.apache.org/jira/browse/SPARK-23633

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -112,12 +113,19 @@ trait CheckAnalysis extends PredicateHelper {
               s"row-based window frame with a single offset: $w")
 
+          case w @ WindowExpression(_: PythonUDF,
+          WindowSpecDefinition(_, _, frame: SpecifiedWindowFrame))
```
indentation :-)

HyukjinKwon commented:
Sorry @icexelloss, I failed to take a close look on this weekend but seems fine in general. Will try it in the coming days.

BTW, I think we should cc @hvanhovell too.

icexelloss committed:
Address PR comments

icexelloss commented:
@HyukjinKwon @viirya Thanks much for the review. It'd be great if @hvanhovell can help review this too!

SparkQA commented:
**[Test build #90007 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90007/testReport)** for PR 21082 at commit [`abfa42b`](https://github.com/apache/spark/commit/abfa42b61344ee43db256dd3d382c41e1868bee1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2774/
Test PASSed.

SparkQA commented:
**[Test build #90007 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90007/testReport)** for PR 21082 at commit [`abfa42b`](https://github.com/apache/spark/commit/abfa42b61344ee43db256dd3d382c41e1868bee1).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

icexelloss committed:
Fix tests

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
**[Test build #90057 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90057/testReport)** for PR 21082 at commit [`875a88e`](https://github.com/apache/spark/commit/875a88ed7a1653d53596825a9232da39dac535c1).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90057/
Test FAILed.

icexelloss commented:
retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2826/
Test PASSed.

gatorsmile commented:
cc @ueshin and @hvanhovell

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90072/
Test FAILed.

icexelloss commented:
Seems like some issue with R build. I will retry later.

hvanhovell commented:
It should be fixed. Let's retry.

hvanhovell commented:
retest this please

HyukjinKwon commented:
retest this please

HyukjinKwon commented:
Oops ..

SparkQA commented:
**[Test build #90143 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90143/testReport)** for PR 21082 at commit [`875a88e`](https://github.com/apache/spark/commit/875a88ed7a1653d53596825a9232da39dac535c1).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2880/
Test FAILed.

SparkQA commented:
**[Test build #90143 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90143/testReport)** for PR 21082 at commit [`875a88e`](https://github.com/apache/spark/commit/875a88ed7a1653d53596825a9232da39dac535c1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90143/
Test PASSed.

icexelloss commented:
Hi @ueshin @hvanhovell can you take a look at this please? Thanks!

BryanCutler commented on python/pyspark/worker.py:
```diff
@@ -128,6 +128,17 @@ def wrapped(*series):
+    def wrapped(*series):
+        import pandas as pd
+        result = f(*series)
+        return pd.Series([result]).repeat(len(series[0]))
```
Just wondering why this needs to be repeated to the length of the series and grouped agg doesn't?

HyukjinKwon replied:
Let's leave a short comment where we are here.

icexelloss replied:
Added comments to describe the function

BryanCutler reviewed (commented):
Looks reasonable to me, but I'm not too familiar with windowing in Spark.  If I understand correctly, this change is a stepping stone to enable bounded windows next?

icexelloss commented:
Gentle ping @hvanhovell @ueshin 

Is there anything else I can do to move forward with this PR?

ueshin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -268,3 +269,38 @@ object PhysicalAggregation {
+    case expr @ logical.Window(windowExpressions, partitionSpec, orderSpec, child) =>
+
+      if (windowExpressions.isEmpty) {
+        throw new AnalysisException(s"Window expression is empty in $expr")
```
Can we reach here?

icexelloss replied:
Window expression should not be empty here so this should not be reached. This is just for safety. Let me add a comment.

icexelloss replied:
I added comments to explain this.

ueshin reviewed (commented):
I'm sorry for the late review.
I left some comments. Thanks!

icexelloss commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -1869,6 +1870,8 @@ class Analyzer(
           case window: WindowExpression => window.windowSpec
         }.distinct
 
+        val windowFunctionType = WindowFunctionType.functionType(expr)
```
This is used for grouping window functions into sql window functions and pandas UDFs and create different logical node for them. The value is used here https://github.com/apache/spark/pull/21082/files#diff-57b3d87be744b7d79a9beacf8e5e5eb2R1886

ueshin replied:
Yeah, I can see it is included in the tuple, but seems like it is used for nothing after that and dropped at https://github.com/apache/spark/pull/21082/files#diff-57b3d87be744b7d79a9beacf8e5e5eb2R1886?

icexelloss replied:
Ah I see. Let me inline this.

icexelloss replied:
Inlined.

ueshin replied:
Oh, now I understand that the long block is for `groupBy`, so the tuple will be the group key, right? I misread the block was for `map` or something. Thanks!

icexelloss commented on python/pyspark/sql/tests.py:
```diff
@@ -5181,6 +5190,235 @@ def test_invalid_args(self):
+            .withColumn('w', lit(1.0))
+
+    @property
+    def python_plus_one(self):
```
I agree those should be shared, but let's do it maybe in a separate PR? Because the refactor will probably touch many test cases and the PR is quite large already..

ueshin replied:
Sure.

icexelloss committed:
Add comments in Physical Window

SparkQA commented:
**[Test build #90996 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90996/testReport)** for PR 21082 at commit [`ed02823`](https://github.com/apache/spark/commit/ed028237d9892e8d1e32607e77dad24ad0252b05).

icexelloss commented:
@ueshin I responded to your comments. Please let me know what you think, thanks!

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/3477/
Test PASSed.

SparkQA commented:
**[Test build #90996 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90996/testReport)** for PR 21082 at commit [`ed02823`](https://github.com/apache/spark/commit/ed028237d9892e8d1e32607e77dad24ad0252b05).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90996/
Test FAILed.

icexelloss committed:
inline window function type

SparkQA commented:
**[Test build #91038 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91038/testReport)** for PR 21082 at commit [`1c6b5d5`](https://github.com/apache/spark/commit/1c6b5d506ae68f132625874444c11380be9eb639).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/3509/
Test PASSed.

SparkQA commented:
**[Test build #91038 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91038/testReport)** for PR 21082 at commit [`1c6b5d5`](https://github.com/apache/spark/commit/1c6b5d506ae68f132625874444c11380be9eb639).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91038/
Test PASSed.

icexelloss commented:
Hey @hvanhovell I think @HyukjinKwon @ueshin @BryanCutler @viirya all reviewed this patch. Could you please take a look :)

icexelloss commented:
Hey @gatorsmile and @cloud-fan sorry to bother you but this I have addressed all comments on the PR and there are no new comments for a while.

Could you please take a look and let me know what to do to move this PR forward? Thank you very much!

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/WindowInPandasExec.scala:
```diff
@@ -0,0 +1,173 @@
+    orderSpec: Seq[SortOrder],
+    child: SparkPlan) extends UnaryExecNode {
+
+  override def output: Seq[Attribute] =
```
Minor comment: should we create a common base class for native Windows and python Windows and ?

icexelloss replied:
I am not sure. Currently python and window are under different package namespaces (execution.python and execution.window). To create a common base class we probably need to refactor the namespace hierarchy somehow and remove some of the `private[python]` and `private[window]`. 

I think we will face the problem when trying to do rolling window with pandas UDF, because we likely want to reuse some of code under `execution.window` package, but I am not sure we should resolve the python and window namespace in this PR.

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -2321,7 +2323,30 @@ def pandas_udf(f=None, returnType=None, functionType=None):
+       >>> @pandas_udf("double", PandasUDFType.GROUPED_AGG)  # doctest: +SKIP
+       ... def mean_udf(v):
+       ...     return v.mean()
+       >>> w = Window.partitionBy('id')
```
Shall we explicitly show unbounded boundaries?

HyukjinKwon commented on python/pyspark/sql/tests.py:
```diff
@@ -5181,6 +5190,235 @@ def test_invalid_args(self):
+
+        result1 = df.withColumn('mean_v', self.pandas_agg_mean_udf(df['v']).over(w)) \
+                    .withColumn('max_v', self.pandas_agg_max_udf(df['v']).over(w)) \
+                    .withColumn('min_w', self.pandas_agg_min_udf(df['w']).over(w)) \
```
Trailing `\`.

HyukjinKwon commented on python/pyspark/sql/tests.py:
```diff
@@ -5181,6 +5190,235 @@ def test_invalid_args(self):
+
+    @property
+    def unpartitioned_window(self):
+        return Window.partitionBy()
```
Shall we test `Window.rowsBetween(Window.unboundedPreceding, Window.unboundedFollowing)` too?

icexelloss replied:
I think we can reply on that `Window.partitionBy()` returns unbounded window here, otherwise there might be too many combinations to test. But I am ok to add the tests for `Window,.partitionBy().rowsBetween(Window.unboundedPreceding, Window.unboundedFollowing)` in addition to the existing ones. WDYT?

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:
```diff
@@ -112,12 +113,19 @@ trait CheckAnalysis extends PredicateHelper {
+          case _ @ WindowExpression(_: PythonUDF,
+            WindowSpecDefinition(_, _, frame: SpecifiedWindowFrame))
+              if !frame.isUnbounded =>
+            failAnalysis(s"Only unbounded window frame is supported with Pandas UDFs.")
```
leading `s` seems not needed.

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala:
```diff
@@ -34,7 +34,12 @@ object PythonUDF {
+      e.asInstanceOf[PythonUDF].evalType == PythonEvalType.SQL_GROUPED_AGG_PANDAS_UDF
+  }
+
+  def isWindowPandasUDF(e: Expression): Boolean = {
```
Hmmmm .. this bit looks a bit odd because the condition is the same but the only the name is different. We should at least need to leave a comment ...

HyukjinKwon replied:
@icexelloss, can we maybe do:

```
def isWindowPandasUDF(e: Expression): = isGroupedAggPandasUDF(e)
```

and explain why they can be same ..

icexelloss replied:
Sounds good

HyukjinKwon commented on python/pyspark/worker.py:
```diff
@@ -128,6 +128,17 @@ def wrapped(*series):
+    def wrapped(*series):
+        import pandas as pd
+        result = f(*series)
+        return pd.Series([result]).repeat(len(series[0]))
```
So, this place is the only place where it's diverted (by repeating?); therefore, needs Windows specific attribute?

icexelloss replied:
Yes - I tried to do this on the Java side but it's tricky and complicated to merging the input row and output of udf if they are not 1-1 mapping. So I ended up doing this..

HyukjinKwon commented:
Hmmm .. please allow me to take another look within few days.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #91499 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91499/testReport)** for PR 21082 at commit [`1c6b5d5`](https://github.com/apache/spark/commit/1c6b5d506ae68f132625874444c11380be9eb639).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/3822/
Test PASSed.

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:
```diff
@@ -424,6 +424,21 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
+      case PhysicalWindow(
+        WindowFunctionType.Python, windowExprs, partitionSpec, orderSpec, child) =>
+        execution.python.WindowInPandasExec(
+          windowExprs, partitionSpec, orderSpec, planLater(child)) :: Nil
```
tiny nit: I would add a newline below

icexelloss replied:
Added

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala:
```diff
@@ -297,6 +297,37 @@ trait WindowFunction extends Expression {
+      case udf: PythonUDF if PythonUDF.isWindowPandasUDF(udf) => Python
+    }
+
+    // Normally a window expression would either have either a SQL window function, a SQL
```
`either have either` ... :-).

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -268,3 +269,40 @@ object PhysicalAggregation {
+ */
+object PhysicalWindow {
+  // windowFunctionType, windowExpression, partitionSpec, orderSpec, child
+  type ReturnType =
```
nit: `private type`

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/WindowInPandasExec.scala:
```diff
@@ -0,0 +1,173 @@
+   * @return the final resulting projection.
+   */
+  private[this] def createResultProjection(expressions: Seq[Expression]): UnsafeProjection = {
+    val references = expressions.zipWithIndex.map{ case (e, i) =>
```
`p{` -> `p {`

HyukjinKwon reviewed (approved):
LGTM

but just for clarification, @icexelloss, do you have a WIP work or plan to support bounded ones too?

SparkQA commented:
**[Test build #91499 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91499/testReport)** for PR 21082 at commit [`1c6b5d5`](https://github.com/apache/spark/commit/1c6b5d506ae68f132625874444c11380be9eb639).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91499/
Test PASSed.

icexelloss commented:
@HyukjinKwon Thanks for the review! I will address the comments shortly.

And yes, I will work on bounded windows on top of this PR.

icexelloss committed:
Address comments

SparkQA commented:
**[Test build #91574 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91574/testReport)** for PR 21082 at commit [`6350408`](https://github.com/apache/spark/commit/63504088f721036d46f58608c353bdb9f186527f).

icexelloss commented:
Thanks @HyukjinKwon for your review! @ueshin Do you want to take another look too?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/3858/
Test PASSed.

HyukjinKwon commented:
Yea, let's leave this open for few more days in case someone has more comments.

SparkQA commented:
**[Test build #91574 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91574/testReport)** for PR 21082 at commit [`6350408`](https://github.com/apache/spark/commit/63504088f721036d46f58608c353bdb9f186527f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91574/
Test PASSed.

ueshin commented:
LGTM.

HyukjinKwon commented:
@icexelloss, mind resolving the conflict?

SparkQA commented:
**[Test build #91718 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91718/testReport)** for PR 21082 at commit [`328b2c4`](https://github.com/apache/spark/commit/328b2c4e09502a66939d47d6967ceea7ceab6c8c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/3944/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/54/
Test PASSed.

SparkQA commented:
**[Test build #91718 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91718/testReport)** for PR 21082 at commit [`328b2c4`](https://github.com/apache/spark/commit/328b2c4e09502a66939d47d6967ceea7ceab6c8c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91718/
Test PASSed.

HyukjinKwon commented:
Merged to master.

icexelloss commented:
Thanks everyone for the review!

adsk2050 commented:
Hello! this is great work! Thank you for contributing. This code will enable to run functions on window, which take in pd.Series -> Any.

I am wondering if GROUPED_MAP pandas UDF as window functions is also in pipeline or not? 
(Basically pd.Series -> pd.Series over Window.) 
For example:
```
from pyspark.sql import functions as F
from pyspark.sql.types import *

def doCoolStuff(df: pd.DataFrame) -> pd.DataFrame:
  events = df["event"].to_list()
  count = 1
  sets = []
  for event in events:
    sets.append(str(count))
    if event=="buy":
      count+=1   
  df["coolStuff"] = pd.Series(data=sets)
  return df

df = spark.createDataFrame(pd.DataFrame([[1, random.choice(list(range(10))), i, random.random()] for i in range(100)], columns=["user_id", "source_id", "epoch_timestamp", "event_prob"]))\
.withColumn("event", F.when(F.col("event_prob")>F.lit(0.9), "buy").otherwise("view"))\
.withColumn("coolStuff", F.lit(""))\
.persist()

doCoolStuffPDUDF = F.pandas_udf(
  f=doCoolStuff,
  returnType=df.schema,
  functionType=F.PandasUDFType.GROUPED_MAP)

df\
.orderBy(F.col("epoch_timestamp"))\
.groupby("user_id", "source_id")\
.apply(doCoolStuffPDUDF)\
.orderBy(F.col("user_id"), F.col("source_id"), F.col("epoch_timestamp").desc())\
.display()
```

This could simplified to:

```
from pyspark.sql import functions as F
from pyspark.sql.types import *
from pyspark.sql.window import Window

def doCoolStuff(events: pd.Series) -> pd.Series:
  count = 1
  sets = []
  for event in events:
    sets.append(str(count))
    if event=="buy":
      count+=1   
  return pd.Series(data=sets)

doCoolStuffPDUDF = F.pandas_udf(
  f=doCoolStuff,
  returnType=StringType(),
  functionType=F.PandasUDFType.GROUPED_MAP)

df = spark.createDataFrame(pd.DataFrame([[1, random.choice(list(range(10))), i, random.random()] for i in range(100)], columns=["user_id", "source_id", "epoch_timestamp", "event_prob"]))\
.withColumn("event", F.when(F.col("event_prob")>F.lit(0.9), "buy").otherwise("view"))\
.withColumn("coolStuff", doCoolStuffPDUDF(F.col("event"))\
                                        .over(Window.partitionBy("user_id", "source_id").orderBy(F.col("epoch_timestamp"))\
.orderBy(F.col("user_id"), F.col("source_id"), F.col("epoch_timestamp").desc())\
.persist()

df.display()
